### DISCUSSION TITLE
Alternative approach for different wait times in `useDebounce` and `useThrottle`

### DISCUSSION THREAD

**Discussion by marcoroth** [stimulus-use/stimulus-use#64] [Ideas]
Hey @adrienpoly.

I saw you introduced configurable wait times for `useDebounce` and `useThrottle` in #59 and #62. Pretty cool!

I just had another idea how we could handle different wait times for different actions. It's a different approach compared to what we saw in the past. The idea is to call the mixin multiple times, but in my opinion this is less cluttery and more expressive.

It also means, that we don't need the `static debounces` array.

```js
export default class extends ApplicationController {
  connect () {
    useDebounce(this.click, { wait: 100 })
    useDebounce(this.fetch, { wait: 500 })
  }

  click () {
    // this function is debounced with a wait time of 100ms.
  }

  fetch () {
    // this function is debounced with a wait time of 500ms.
  }

  instant () {
    // this function is not debounced.
  }
}
```

It's just an idea for a different approach that I wanted to share with you to get your take on it. Would also be the first time that we don't pass an actual controller to a mixin. I don't know how it would affect the consistency of the overall api, but maybe we can also support both for even more flexibility.

**Comment by adrienpoly** [comment]
Hey 

Thanks for your feedback, my initial though was to keep a similar API as `useMemo` but I agree that the final API is a bit different from what we are used to with `debouncers`.

I published a first beta version that had fixed time for all function. The API is simple with this static array of function name to debounce.
Then realized I would need per function setting for one project to fully removed lodash. This was an edge case as most of the time I am using a single wait time for all debounced function.

@julianrubisch any thought on this as for me ultimately what is important is to **be consistent with `useMemo`**

To get you up to speed without looking in 4 PRs the current API is: 

```js
export default class extends ApplicationController {
  static debounces = ["click", { name: "fetch", wait: 500 }]

  connect () {
    useDebounce(this)
  }

  click () {
    // this function is debounced with a wait time of 200ms (default).
  }

  fetch () {
    // this function is debounced with a wait time of 500ms.
  }

  instant () {
    // this function is not debounced.
  }
}
```

other alternatives 

```js
connect () {
    useDebounce(["click", { name: "fetch", wait: 500 }])
  }
```

```js
connect () {
    useDebounce([this.click, { function: this.fetch, wait: 500 }])
  }
```

**Comment by julianrubisch** [comment]
Hmmm yeah I would probably go with option 1. This mix of plain values with objects in an array is also consistent with common configs such as tailwind etc.

**Comment by adrienpoly** [comment]
Ok I had some thoughts over this and discussed it around

The API proposed by @marcoroth and/or my suggested alternative usually makes more sens for newcomers into a code base as it is very explicit.  
One comment I had that I found interesting is that static arrays (ala targets) are more global to the controller and as it is currently it makes it strange to has `useDebounce` on one line and the functions to debounce at another line… Definitely makes sens.

## Best of both world

#### Using a global ApplicationController

We can keep the current API but recommend it when there is a global `ApplicationController` that instantiate the useDebounce behaviour and then the children controllers can simply debounce functions by defining the `static debounces`

```js
// application_controller.js
export default class extends Controller {

  connect () {
    useDebounce(this)
  }
}
```

```js
// some_controller.js
export default class extends ApplicationController {
  static debounces = ["click", { name: "fetch", wait: 500 }]

  click () {
    // this function is debounced with a wait time of 200ms (default).
  }

  fetch () {
    // this function is debounced with a wait time of 500ms.
  }

  instant () {
    // this function is not debounced.
  }
}
```

#### Per controller definition

When not using a global `ApplicationController` then locally we can debounce (throttle, memo) function with either of those possibilities:

```js
// most basic with default wait time
useDebounce(this.click)
useDebounce(this.fetch)

// default wait time for an array of functions
useDebounce([this.click, this.fetch])

// array of functions with ddefault and custom wait time
useDebounce([this.click, { function: this.fetch, wait: 500 }])

```
